File: festival-web/src/main/java/cn/bdqfork/web/proxy/VerticleProxyHandler.java
package cn.bdqfork.web.proxy;

import cn.bdqfork.web.context.service.HessianMessageCodec;
import cn.bdqfork.web.context.service.MethodInvocation;
import cn.bdqfork.web.util.EventBusUtils;
import io.vertx.core.eventbus.DeliveryOptions;
import io.vertx.reactivex.core.Vertx;
import io.vertx.reactivex.core.eventbus.EventBus;
import io.vertx.reactivex.core.eventbus.Message;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

/**
 * @author bdq
 * @since 2020/1/26
 */
public class VerticleProxyHandler implements InvocationHandler {
    private DeliveryOptions options;
    private Vertx vertx;
    private Class<?> targetClass;

    public VerticleProxyHandler(Vertx vertx, Class<?> targetClass) {
        this.vertx = vertx;
        this.targetClass = targetClass;
        this.options = new DeliveryOptions();
        this.options.setCodecName(HessianMessageCodec.NAME);
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
        MethodInvocation methodInvocation = new MethodInvocation(method, args);
        EventBus eventBus = vertx.eventBus();
        String address = EventBusUtils.getAddress(targetClass);
        return eventBus.rxRequest(address, methodInvocation, options)
                .toFlowable()
                .map(Message::body);
    }
}

File: festival-kotlin-web/src/main/kotlin/cn/bdqfork/kotlin/web/service/VerticleProxyHandler.kt
package cn.bdqfork.kotlin.web.service

import cn.bdqfork.kotlin.web.util.EventBusUtils
import io.vertx.core.Vertx
import io.vertx.core.eventbus.DeliveryOptions
import io.vertx.core.eventbus.Message
import io.vertx.kotlin.coroutines.awaitResult
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

/**
 * @author bdq
 * @since 2020/1/26
 */
class VerticleProxyHandler(private val vertx: Vertx, private val targetClass: Class<*>) : InvocationHandler {
    private val options: DeliveryOptions = DeliveryOptions()
    @Throws(Throwable::class)
    override fun invoke(proxy: Any, method: Method, args: Array<Any?>?): Any? = runBlocking {
        val methodInvocation = MethodInvocation(method, args ?: emptyArray())
        val address = EventBusUtils.getAddress(targetClass)
        return@runBlocking GlobalScope.async {
            val message = awaitResult<Message<Any>> { h -> vertx.eventBus().request(address, methodInvocation, options, h) }
            val result = message.body()
            if (result is Throwable) {
                throw result
            }
            return@async result
        }.await()
    }

    init {
        options.codecName = JsonMessageCodec.NAME
    }
}